native/membrane_videocompositor/src/elixir_bridge/scene_validation.rs

```diff
@@ -0,0 +1,392 @@
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        if objects.is_empty() {
+            return Err(SceneParsingError::UndefinedName(final_object_name.clone()));
+        }
+
+        let mut names = HashSet::new();
+        for (name, _) in objects {
+            if names.contains(name) {
+                return Err(SceneParsingError::DuplicateNames(name.clone()));
+            }
+
+            names.insert(name);
+        }
+
+        for (_, object) in objects {
+            for name in object.mentioned_names() {
+                if !names.contains(name) {
+                    return Err(SceneParsingError::UndefinedName(name.clone()));
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    fn check_for_unused_objects(
+        objects: &[(ObjectName, Object)],
+        final_object_name: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        let defined_names = objects.iter().map(|(name, _)| name).collect::<HashSet<_>>();
+
+        let used_names = objects
+            .iter()
+            .flat_map(|(_, object)| object.mentioned_names())
+            .collect::<HashSet<_>>();
+
+        let difference = defined_names.difference(&used_names).collect::<Vec<_>>();
+
+        if difference.is_empty() {
+            return Err(SceneParsingError::CycleDetected);
+        }
+
+        if *difference[0] != final_object_name {
+            return Err(SceneParsingError::UnusedObject((*difference[0]).clone()));
+        }
+
+        if difference.len() > 1 {
+            return Err(SceneParsingError::UnusedObject((*difference[1]).clone()));
+        }
+
+        Ok(())
+    }
+
+    /// returns true if a cycle exists in the scene graph
+    fn contains_cycle(
+        objects: &HashMap<&ObjectName, &Object>,
+        final_object: &ObjectName,
+    ) -> Result<(), SceneParsingError> {
+        enum NodeState {
+            BeingVisited,
+            Visited,
+        }
+
+        fn visit(
+            name: &ObjectName,
+            objects: &HashMap<&ObjectName, &Object>,
+            visited: &mut HashMap<ObjectName, NodeState>,
+        ) -> Result<(), SceneParsingError> {
+            match visited.get(name) {
+                Some(NodeState::BeingVisited) => return Err(SceneParsingError::CycleDetected),
+                Some(NodeState::Visited) => return Ok(()),
+                _ => {}
+            }
+
+            visited.insert(name.clone(), NodeState::BeingVisited);
+
+            match &objects[name] {
+                Object::Video(_) => Ok(()),
+                Object::Texture(Texture { input, .. }) => visit(input, objects, visited),
+                Object::Layout(Layout { inputs, .. }) => inputs
+                    .values()
+                    .map(|name| visit(name, objects, visited))
+                    .collect::<Result<Vec<()>, SceneParsingError>>()
+                    .map(|_| ()),
+            }?;
```

Comment:
This will silently break if we change how we define neighbor nodes. I think that a separate function like `get_previous_nodes` would make it harder to forget to update that.